File: Go/Day18.go
package main

import AH "./adventhelper"

import (
	"regexp"
	"strconv"
	"strings"
)

// a1[*/+]a2[*/+]a3[*/+]a4[*/+]a5[*/+]a6
// resolves operators left to right
func parseLR(exp string) (string) {
	prevExp := ""
	re := regexp.MustCompile("[0-9]+[\\*\\+][0-9]+")
	for {
		if exp == prevExp {
			break
		}
		prevExp = exp
		found := re.FindString(exp)
		if found != "" {
			newValue := ""
			if AH.ContainsChar(found, '*') {
				parts := strings.Split(found, "*")
				lhs, _ := strconv.Atoi(parts[0])
				rhs, _ := strconv.Atoi(parts[1])
				newValue = strconv.Itoa(lhs * rhs)
			}

			if AH.ContainsChar(found, '+') {
				parts := strings.Split(found, "+")
				lhs, _ := strconv.Atoi(parts[0])
				rhs, _ := strconv.Atoi(parts[1])
				newValue = strconv.Itoa(lhs + rhs)
			}

			exp = strings.Replace(exp, found, newValue, 1)
		}
	}

	return exp
}

// a1[*/+]a2[*/+]a3[*/+]a4[*/+]a5[*/+]a6
// resolves all additions the multiplications
func parsePM(exp string) (string) {
	prevExp := ""
	reAdd := regexp.MustCompile("[0-9]+[\\+][0-9]+")
	reMul := regexp.MustCompile("[0-9]+[\\*][0-9]+")

	for {
		if (exp == prevExp) {
			break
		}
		prevExp = exp

		// look for an addition term
		found := reAdd.FindString(exp)
		if found != "" {
			parts := strings.Split(found, "+")
			lhs, _ := strconv.Atoi(parts[0])
			rhs, _ := strconv.Atoi(parts[1])
			newValue := strconv.Itoa(lhs + rhs)

			exp = strings.Replace(exp, found, newValue, 1)
			continue
		}
		// if we get here we haven't found an addition term,
		// so look for a multiplcation
		found = reMul.FindString(exp)
		if found != "" {
			parts := strings.Split(found, "*")
			lhs, _ := strconv.Atoi(parts[0])
			rhs, _ := strconv.Atoi(parts[1])
			newValue := strconv.Itoa(lhs * rhs)

			exp = strings.Replace(exp, found, newValue, 1)
			continue
		}

		return exp
	}

	return exp
}

// find a bracket that does not conatin any other brackets
func parseBracket(exp string, lr bool) (string) {
	prevExp := ""
	re := regexp.MustCompile("\\([0-9+*]+\\)")
	for {
		if exp == prevExp {
			break
		}
		prevExp = exp
		found := re.FindString(exp)
		// have we found a bracket with no nested brackets inside?
		if found != "" {
			// if so reduce it, based on the rule given in the instructions
			newValue := ""
			if lr {
				newValue = parseLR(AH.TrimLastRune(AH.TrimFirstRune(found)))
			} else {
				newValue = parsePM(AH.TrimLastRune(AH.TrimFirstRune(found)))
			}

			exp = strings.Replace(exp, found, newValue, 1)
		}
	}
	return exp
}

func Part1(exp string) (value int) {
	value, _ = strconv.Atoi(parseLR(parseBracket(exp, true)))
	return
}

func Part2(exp string) (value int) {
	value, _ = strconv.Atoi(parsePM(parseBracket(exp, false)))
	return
}

func main() {
	ss, _ := AH.ReadStrFile("../input/input18.txt")
	expressions := []string{}
	for _, s := range ss {
		temp := strings.Replace(s, " ", "", -1) // remove the spaces!
		expressions = append(expressions, temp)
	}

	total1, total2 := 0, 0
	for _, e := range expressions {
		total1 += Part1(e)
		total2 += Part2(e)
	}

	AH.PrintSoln(18, total1, total2)

	return
}

File: C++/Day18.cpp
#include "AH.h"

namespace Day18
{

	std::string parseLR(const std::string s)
	{
		std::string exp = s;
		std::string prevExp = "";
		auto re = std::regex("[0-9]+[\\*\\+][0-9]+");
		while (exp != prevExp)
		{
			prevExp = exp;
			std::smatch m;
			if (regex_search(exp, m, re))
			{
				std::string found = m[0];

				std::string newValue = "";
				if (found.find('*') < found.length())
				{
					auto parts = AH::Split(found, '*');
					int64_t lhs = std::stoll(parts[0]);
					int64_t rhs = std::stoll(parts[1]);
					newValue = std::to_string(lhs * rhs);
				}

				if (found.find('+') < found.length())
				{
					auto parts = AH::Split(found, '+');
					int64_t lhs = std::stoll(parts[0]);
					int64_t rhs = std::stoll(parts[1]);
					newValue = std::to_string(lhs + rhs);
				}

				exp = std::regex_replace(exp, re, newValue,
					                       std::regex_constants::format_first_only);
			}
		}

		return exp;
	}

	std::string parsePM(const std::string s)
	{
		std::string exp = s;
		std::string prevExp = "";
		auto reAdd = std::regex("[0-9]+[\\+][0-9]+");
		auto reMul = std::regex("[0-9]+[\\*][0-9]+");
		while (exp != prevExp)
		{
			prevExp = exp;
			std::smatch m;
			if (regex_search(exp, m, reAdd))
			{
 				std::string found = m[0];
				auto parts = AH::Split(found, '+');
				int64_t lhs = std::stoll(parts[0]);
				int64_t rhs = std::stoll(parts[1]);
				std::string newValue = std::to_string(lhs + rhs);
				exp = std::regex_replace(exp, reAdd, newValue,
					                       std::regex_constants::format_first_only);
				continue;
			}

			if (regex_search(exp, m, reMul))
			{
				std::string found = m[0];
				auto parts = AH::Split(found, '*');
				int64_t lhs = std::stoll(parts[0]);
				int64_t rhs = std::stoll(parts[1]);
				std::string newValue = std::to_string(lhs * rhs);
				exp = std::regex_replace(exp, reMul, newValue,
					                       std::regex_constants::format_first_only);
				continue;
			}
		}

		return exp;
	}

	std::string parseBracket(const std::string s, const bool lr)
	{
		std::string exp = s;
		std::string prevExp = "";
		auto re = std::regex("\\([0-9+*]+\\)");

		while (exp != prevExp)
		{
			prevExp = exp;
			std::smatch m;
			if (regex_search(exp, m, re))
			{
				std::string found = m[0];
				found = found.substr(1, found.size() - 2);

				std::string newValue = "";
				if (lr)
				{
					newValue = parseLR(found);
				}
				else
				{
					newValue = parsePM(found);
				}
				exp = std::regex_replace(exp, re, newValue,
					                       std::regex_constants::format_first_only);
			}

		}

		return exp;
	}

	int64_t Evaluate(const std::string exp, const bool part1)
	{
		auto brackets = parseBracket(exp, part1);

		if (part1)
			return std::stoll(parseLR(brackets));
		else
			return std::stoll(parsePM(brackets));
	}



	int Run(const std::string& filename)
	{
		auto inputLines = AH::ReadTextFile(filename);
		std::vector<std::string> expressions;
		for ( auto& e : inputLines)
		{
			e.erase(remove_if(e.begin(), e.end(), isspace), e.end());
			expressions.push_back(e);
		}

		int64_t part1 = 0, part2 = 0;
		for ( auto e : expressions )
		{
			part1 += Evaluate(e, true);
			part2 += Evaluate(e, false);
		}

		AH::PrintSoln(18, part1, part2);

		return 0;
	}

}
